### ISSUE TITLE
tf.experimental.numpy.heaviside` and `tf.keras.ops.heaviside` return finite values for NaN input instead of propagating NaN

### ISSUE THREAD

**Issue by dingjiawei65** [tensorflow/tensorflow#127819]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

tf2.20.0

### Custom code

Yes

### OS platform and distribution

Linux 6.8.0-138-generic x86_64

### Python version

3.11.14

### CUDA/cuDNN version

CUDA  12.5.1

### Current behavior?

text
NumPy:                              [nan, 1.0, 0.0]
tf.experimental.numpy.heaviside:    [0.0, 1.0, 0.0]
tf.keras.ops.heaviside:             [0.0, 1.0, 0.0]
```

Finite controls match NumPy:

```text
input [2.0, 0.0, -2.0], values [0.5, 0.5, 0.5]
expected/output: [1.0, 0.5, 0.0]


For NumPy compatibility, `tf.experimental.numpy.heaviside` should preserve `NaN` in the first input, matching:

```text
np.heaviside([nan, 0.0, -1.0], [1.0, 1.0, 1.0]) -> [nan, 1.0, 0.0]
```

If `tf.keras.ops.heaviside` intentionally follows a different contract, it would help to document the NaN behavior.

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf

print("TensorFlow:", tf.__version__)
print("Keras:", tf.keras.__version__)
print("NumPy:", np.__version__)

x = np.array([np.nan, 0.0, -1.0], dtype=np.float32)
values = np.array([1.0, 1.0, 1.0], dtype=np.float32)

print("NumPy:", np.heaviside(x, values), "isnan:", np.isnan(np.heaviside(x, values)).tolist())

for name, fn in [
    ("tf.experimental.numpy.heaviside", tf.experimental.numpy.heaviside),
    ("tf.keras.ops.heaviside", tf.keras.ops.heaviside),
]:
    result = fn(tf.constant(x), tf.constant(values))
    print(name, result.numpy(), "isnan:", np.isnan(result.numpy()).tolist())

finite_x = np.array([2.0, 0.0, -2.0], dtype=np.float32)
finite_values = np.array([0.5, 0.5, 0.5], dtype=np.float32)
print("finite NumPy:", np.heaviside(finite_x, finite_values))
for name, fn in [
    ("tf.experimental.numpy.heaviside", tf.experimental.numpy.heaviside),
    ("tf.keras.ops.heaviside", tf.keras.ops.heaviside),
]:
    print("finite", name, fn(tf.constant(finite_x), tf.constant(finite_values)).numpy())
```

### Relevant log output

```shell

```

**Comment by Kayyuri** [comment]
Hi @dingjiawei65,  Thank you for reporting this issue!

You've clearly demonstrated that the `tf.experimental.numpy.heaviside` and `tf.keras.ops.heaviside` returns `0.0` for `NaN` inputs instead of propagating `NaN`, which is inconsistent with `numpy.heaviside` behavior. 

We’ll investigate the issue on our end and provide an update.

**Comment by Phoebus-Liu** [comment]
Hi @Kayyuri, thank you for triaging this. I've opened #127890 to fix the TensorFlow-side regression in `tf.experimental.numpy.heaviside`.

**Root cause.** The implementation chooses the result with `where_v2(x1 < 0, 0, where_v2(x1 > 0, 1, x2))`. TensorFlow's relational operators return `False` for NaN inputs, so for `x1 = nan` neither branch is taken and the value silently falls through to `x2` — e.g. `heaviside([nan, 0, -1, 1], [1, 1, 1, 1])` returned `[1, 1, 0, 1]` instead of `[nan, 1, 0, 1]`. The fix masks out NaN elements of the first argument and returns NaN for them, matching NumPy. Integer dtypes cannot hold NaN, so the mask is only applied to floating-point inputs (integer results are still promoted to floating point as before).

**Scope.** This covers `tf.experimental.numpy.heaviside` in this repository. `tf.keras.ops.heaviside` lives in the separate `keras-team/keras` package, so I left it out of this PR to keep the change focused — happy to open a companion keras PR if you'd like, or to adjust the approach (e.g. also touch the keras path here) if you prefer a single issue-level fix.

**Verification.** On CPU with TF 2.21.0 and nightly 2.22.0.dev, the example above now matches `np.heaviside`. I added three regression tests (`testHeaviside`, `testHeavisideNaNPropagation`, `testHeavisideIntegerInputs`); reverting the change makes `testHeavisideNaNPropagation` fail, confirming the test locks the bug.

Please let me know if you'd like any changes to the implementation or the split between the TF and keras packages. Thank you for maintaining this issue.

**Comment by Kayyuri** [comment]
Hi @dingjiawei65,
A fix for this issue has been proposed in this [PR](https://github.com/tensorflow/tensorflow/pull/127890). Once it is merged, this issue will be resolved.

Big thanks to @Phoebus-Liu  for raising the PR to address this issue. We truly appreciate your initiative and the effort in contributing the fix.

**Comment by Phoebus-Liu** [comment]
Hi @Kayyuri,

Thank you — my pleasure. PR #127890 has been approved and is queued for merge, so this issue should be resolved once it lands.

One note for the record: that PR covers the TensorFlow-side `tf.experimental.numpy.heaviside` only; the Keras path would need a separate change if the same behavior is desired there.

Thanks again for the triage!
